src/workflows/annotation/scgpt_integration_knn/main.nf

```diff
@@ -0,0 +1,194 @@
+                "input_var_key": "input_var_gene_names"
+            ],
+            args: [
+                "output_var_key": "gene_symbols"
+            ],
+            toState: [
+                "input": "output"
+            ]
+        )
+        | copy_var.run(
+            fromState: [
+                "input": "reference",
+                "modality": "modality",
+                "input_var_key": "reference_var_gene_names"
+            ],
+            args: [
+                "output_var_key": "gene_symbols"
+            ],
+            toState: [
+                "reference": "output"
+            ]
+        )
+        // Concatenate query and reference datasets prior to integration
+        | concatenate_h5mu.run(
+            fromState: { id, state -> [
+                "input": [state.input, state.reference]
+                ]
+            },
+            args: [
+                "input_id": ["query", "reference"],
+                "other_axis_mode": "move"
+            ],
+            toState: ["input": "output"]
+            )
+        | view {"After concatenation: $it"}
+        // Run scgpt integration with leiden clustering
+        | scgpt_leiden_workflow.run(
+            fromState: { id, state ->
+            [
+                "id": id,
+                "input": state.input,
+                "modality": state.modality,
+                "input_layer": state.input_layer,
+                "model": state.model,
+                "model_vocab": state.model_vocab,
+                "model_config": state.model_config,
+                "finetuned_checkpoints_key": state.finetuned_checkpoints_key,
+                "obsm_integrated": state.output_obsm_integrated,
+                "pad_token": state.pad_token,
+                "pad_value": state.pad_value,
+                "n_hvg": state.n_hvg,
+                "max_seq_len": state.max_seq_len,
+                "dsbn": state.dsbn,
+                "batch_size": state.batch_size,
+                "n_input_bins": state.n_input_bins,
+                "seed": state.seed,
+                "leiden_resolution": state.leiden_resolution,
+            ]},
+            args: [
+                "obs_batch_label": "batch_label",
+                "var_gene_names": "gene_symbols"
+            ],
+            toState: ["input": "output"]
+            )
+        | view {"After integration: $it"}
+        // Split integrated dataset back into a separate reference and query dataset
+        | split_h5mu.run(
+            fromState: [
+                "input": "input",
+                "modality": "modality"
+            ],
+            args: [
+                "obs_feature": "dataset",
+                "output_files": "sample_files.csv",
+                "drop_obs_nan": "true",
+                "output": "ref_query"
+            ],
+            toState: [ 
+                "output": "output", 
+                "output_files": "output_files" 
+            ],
+            auto: [ publish: true ]
```

Comment:
Viash handles the publishing at the end of a workflow now, so no need to do the publishing per-component. So what you need to do is add all the outputs to the viash config and keep the entries in the state until the end of workflow. This way Viash can also check if every required output is present.
```suggestion
```

If you do not need these outputs in the publish dir (only the output h5mu), you can just drop this line.